File: tests/statistical_tests/spot_checks/anova-two-way.test.ts
import { twoWayAnova } from "../../../src/dataframe/ts/stats/statistical-tests/anova.ts";

Deno.test("Two-Way ANOVA Spot Check", () => {
  // Data from R test - 2x2 factorial design
  // Factor A: Treatment (Control vs Treatment)
  // Factor B: Time (Before vs After)
  // 6 observations per cell

  // Cell A1B1: Control, Before
  const groupA1B1 = [12.3, 15.7, 18.2, 14.8, 16.1, 13.9];

  // Cell A1B2: Control, After
  const groupA1B2 = [22.1, 25.4, 28.6, 24.3, 26.8, 23.7];

  // Cell A2B1: Treatment, Before
  const groupA2B1 = [11.8, 14.2, 17.5, 13.9, 15.6, 12.4];

  // Cell A2B2: Treatment, After
  const groupA2B2 = [28.5, 31.2, 34.8, 30.1, 32.7, 29.3];

  // Create 3D data structure for two-way ANOVA
  const data = [
    [groupA1B1, groupA1B2], // Control: Before, After
    [groupA2B1, groupA2B2], // Treatment: Before, After
  ];

  // Perform two-way ANOVA
  const result = twoWayAnova({
    data: data,
    alpha: 0.05,
  });

  // Display comprehensive results
  console.log("=== TYPESCRIPT TWO-WAY ANOVA RESULT ===");

  // Basic ANOVA table
  console.log("ANOVA Table:");
  console.log(
    `Factor A (Treatment): F(${result.factor_a.degrees_of_freedom},${result.df_error}) = ${
      result.factor_a.test_statistic.value.toFixed(4)
    }, p = ${result.factor_a.p_value.toFixed(5)}`,
  );
  console.log(
    `Factor B (Time): F(${result.factor_b.degrees_of_freedom},${result.df_error}) = ${
      result.factor_b.test_statistic.value.toFixed(4)
    }, p = ${result.factor_b.p_value.toFixed(5)}`,
  );
  console.log(
    `Interaction (A×B): F(${result.interaction.degrees_of_freedom},${result.df_error}) = ${
      result.interaction.test_statistic.value.toFixed(4)
    }, p = ${result.interaction.p_value.toFixed(5)}`,
  );

  // Sum of squares
  console.log(
    `Sum of Squares - Factor A: ${
      result.factor_a.sum_of_squares.toFixed(4)
    }, Factor B: ${result.factor_b.sum_of_squares.toFixed(4)}, Interaction: ${
      result.interaction.sum_of_squares.toFixed(4)
    }`,
  );
  console.log(
    `Residual SS: ${result.anova_table[3].ss.toFixed(4)}, Total SS: ${
      result.anova_table[4].ss.toFixed(4)
    }`,
  );

  // Mean squares
  console.log(
    `Mean Squares - Factor A: ${
      result.factor_a.mean_square.toFixed(4)
    }, Factor B: ${result.factor_b.mean_square.toFixed(4)}, Interaction: ${
      result.interaction.mean_square.toFixed(4)
    }`,
  );
  console.log(`Residual MS: ${result.ms_error.toFixed(4)}`);

  // Effect sizes (from ANOVA table components)
  console.log(
    `Eta-squared - Factor A: ${
      result.anova_table[0].eta_squared?.toFixed(4) || "N/A"
    }, Factor B: ${
      result.anova_table[1].eta_squared?.toFixed(4) || "N/A"
    }, Interaction: ${result.anova_table[2].eta_squared?.toFixed(4) || "N/A"}`,
  );
  console.log(
    `Partial Eta-squared - Factor A: ${
      result.anova_table[0].partial_eta_squared?.toFixed(4) || "N/A"
    }, Factor B: ${
      result.anova_table[1].partial_eta_squared?.toFixed(4) || "N/A"
    }, Interaction: ${
      result.anova_table[2].partial_eta_squared?.toFixed(4) || "N/A"
    }`,
  );
  console.log(
    `Omega-squared - Factor A: ${
      result.anova_table[0].omega_squared?.toFixed(4) || "N/A"
    }, Factor B: ${
      result.anova_table[1].omega_squared?.toFixed(4) || "N/A"
    }, Interaction: ${
      result.anova_table[2].omega_squared?.toFixed(4) || "N/A"
    }`,
  );

  // Model fit
  console.log(`R-squared: ${result.r_squared.toFixed(4)}`);
  console.log(
    `Sample size: ${result.sample_size} | Total df: ${result.df_total} | Error df: ${result.df_error}`,
  );
  console.log(`Grand mean: ${result.grand_mean.toFixed(4)}`);

  // Cell means and standard deviations
  console.log("Cell means:");
  console.log(
    `  Control-Before: ${result.sample_means[0].toFixed(4)} (SD: ${
      result.sample_std_devs[0].toFixed(4)
    })`,
  );
  console.log(
    `  Control-After: ${result.sample_means[1].toFixed(4)} (SD: ${
      result.sample_std_devs[1].toFixed(4)
    })`,
  );
  console.log(
    `  Treatment-Before: ${result.sample_means[2].toFixed(4)} (SD: ${
      result.sample_std_devs[2].toFixed(4)
    })`,
  );
  console.log(
    `  Treatment-After: ${result.sample_means[3].toFixed(4)} (SD: ${
      result.sample_std_devs[3].toFixed(4)
    })`,
  );

  // Degrees of freedom breakdown
  console.log(
    `Degrees of freedom - Factor A: ${result.factor_a.degrees_of_freedom}, Factor B: ${result.factor_b.degrees_of_freedom}, Interaction: ${result.interaction.degrees_of_freedom}, Error: ${result.df_error}, Total: ${result.df_total}`,
  );

  // ANOVA table components
  console.log("ANOVA Table Components:");
  result.anova_table.forEach((component) => {
    console.log(
      `  ${component.component}: SS=${
        component.ss.toFixed(4)
      }, df=${component.df}, MS=${component.ms?.toFixed(4) || "N/A"}, F=${
        component.f_statistic?.toFixed(4) || "N/A"
      }, p=${component.p_value?.toFixed(5) || "N/A"}`,
    );
  });

  // Test name and alpha
  console.log(`Test name: ${result.test_name}`);
  console.log(`Alpha level: ${result.alpha}`);

  console.log("=== END TYPESCRIPT TWO-WAY ANOVA ===");
});

File: tests/statistical_tests/spot_checks/anova-two-way.test.R
# Two-way ANOVA spot check test data
# Create a 2x2 factorial design with known effects

# Factor A: Treatment (Control vs Treatment)
# Factor B: Time (Before vs After)
# 2x2 design with 6 observations per cell

# Cell A1B1: Control, Before
groupA1B1 <- c(12.3, 15.7, 18.2, 14.8, 16.1, 13.9)

# Cell A1B2: Control, After  
groupA1B2 <- c(22.1, 25.4, 28.6, 24.3, 26.8, 23.7)

# Cell A2B1: Treatment, Before
groupA2B1 <- c(11.8, 14.2, 17.5, 13.9, 15.6, 12.4)

# Cell A2B2: Treatment, After
groupA2B2 <- c(28.5, 31.2, 34.8, 30.1, 32.7, 29.3)

# Combine all data
y <- c(groupA1B1, groupA1B2, groupA2B1, groupA2B2)

# Create factor variables
factorA <- factor(rep(c("Control", "Treatment"), each = 12))
factorB <- factor(rep(c("Before", "After", "Before", "After"), each = 6))

# Create data frame
df <- data.frame(
  y = y,
  factorA = factorA,
  factorB = factorB
)

# Perform two-way ANOVA
model <- aov(y ~ factorA * factorB, data = df)
summary_result <- summary(model)

# Extract ANOVA table components
anova_table <- summary_result[[1]]

# Extract individual test results
factorA_ss <- anova_table[1, "Sum Sq"]
factorA_df <- anova_table[1, "Df"]
factorA_ms <- anova_table[1, "Mean Sq"]
factorA_f <- anova_table[1, "F value"]
factorA_p <- anova_table[1, "Pr(>F)"]

factorB_ss <- anova_table[2, "Sum Sq"]
factorB_df <- anova_table[2, "Df"]
factorB_ms <- anova_table[2, "Mean Sq"]
factorB_f <- anova_table[2, "F value"]
factorB_p <- anova_table[2, "Pr(>F)"]

interaction_ss <- anova_table[3, "Sum Sq"]
interaction_df <- anova_table[3, "Df"]
interaction_ms <- anova_table[3, "Mean Sq"]
interaction_f <- anova_table[3, "F value"]
interaction_p <- anova_table[3, "Pr(>F)"]

residual_ss <- anova_table[4, "Sum Sq"]
residual_df <- anova_table[4, "Df"]
residual_ms <- anova_table[4, "Mean Sq"]

# Calculate total sum of squares
total_ss <- sum(anova_table[, "Sum Sq"])
total_df <- sum(anova_table[, "Df"])

# Calculate R-squared
r_squared <- 1 - (residual_ss / total_ss)
adjusted_r_squared <- 1 - (residual_ms / (total_ss / total_df))

# Calculate cell means
cell_means <- tapply(y, list(factorA, factorB), mean)
grand_mean <- mean(y)

# Calculate cell standard deviations
cell_sds <- tapply(y, list(factorA, factorB), sd)

# Calculate effect sizes (eta-squared)
eta_squared_factorA <- factorA_ss / total_ss
eta_squared_factorB <- factorB_ss / total_ss
eta_squared_interaction <- interaction_ss / total_ss

# Calculate partial eta-squared
partial_eta_squared_factorA <- factorA_ss / (factorA_ss + residual_ss)
partial_eta_squared_factorB <- factorB_ss / (factorB_ss + residual_ss)
partial_eta_squared_interaction <- interaction_ss / (interaction_ss + residual_ss)

# Calculate omega-squared (more conservative effect size)
omega_squared_factorA <- (factorA_ss - factorA_df * residual_ms) / (total_ss + residual_ms)
omega_squared_factorB <- (factorB_ss - factorB_df * residual_ms) / (total_ss + residual_ms)
omega_squared_interaction <- (interaction_ss - interaction_df * residual_ms) / (total_ss + residual_ms)

# Extract model information
model_info <- summary.lm(model)
coefficients <- model_info$coefficients
residual_se <- model_info$sigma
r_squared_lm <- model_info$r.squared
adj_r_squared_lm <- model_info$adj.r.squared

# Extract F-statistic for overall model
f_statistic <- model_info$fstatistic[1]
f_df1 <- model_info$fstatistic[2]
f_df2 <- model_info$fstatistic[3]
f_p_value <- pf(f_statistic, f_df1, f_df2, lower.tail = FALSE)

# Extract degrees of freedom
df_total <- length(y) - 1
df_error <- residual_df

# Calculate sample size
sample_size <- length(y)

# Output comprehensive results
cat("=== R TWO-WAY ANOVA RESULT ===\n")

# Basic ANOVA table
cat("ANOVA Table:\n")
cat(sprintf("Factor A (Treatment): F(%.0f,%.0f) = %.4f, p = %.5f\n", 
    factorA_df, residual_df, factorA_f, factorA_p))
cat(sprintf("Factor B (Time): F(%.0f,%.0f) = %.4f, p = %.5f\n", 
    factorB_df, residual_df, factorB_f, factorB_p))
cat(sprintf("Interaction (A×B): F(%.0f,%.0f) = %.4f, p = %.5f\n", 
    interaction_df, residual_df, interaction_f, interaction_p))

# Sum of squares
cat(sprintf("Sum of Squares - Factor A: %.4f, Factor B: %.4f, Interaction: %.4f, Residual: %.4f, Total: %.4f\n",
    factorA_ss, factorB_ss, interaction_ss, residual_ss, total_ss))

# Mean squares
cat(sprintf("Mean Squares - Factor A: %.4f, Factor B: %.4f, Interaction: %.4f, Residual: %.4f\n",
    factorA_ms, factorB_ms, interaction_ms, residual_ms))

# Effect sizes
cat(sprintf("Eta-squared - Factor A: %.4f, Factor B: %.4f, Interaction: %.4f\n",
    eta_squared_factorA, eta_squared_factorB, eta_squared_interaction))

cat(sprintf("Partial Eta-squared - Factor A: %.4f, Factor B: %.4f, Interaction: %.4f\n",
    partial_eta_squared_factorA, partial_eta_squared_factorB, partial_eta_squared_interaction))

cat(sprintf("Omega-squared - Factor A: %.4f, Factor B: %.4f, Interaction: %.4f\n",
    omega_squared_factorA, omega_squared_factorB, omega_squared_interaction))

# Model fit
cat(sprintf("R-squared: %.4f | Adjusted R-squared: %.4f\n", r_squared, adjusted_r_squared))
cat(sprintf("Residual Standard Error: %.4f\n", residual_se))
cat(sprintf("Overall F(%.0f,%.0f) = %.4f, p = %.5f\n", f_df1, f_df2, f_statistic, f_p_value))

# Sample information
cat(sprintf("Sample size: %d | Total df: %d | Error df: %d\n", sample_size, df_total, df_error))
cat(sprintf("Grand mean: %.4f\n", grand_mean))

# Cell means and standard deviations
cat("Cell means:\n")
cat(sprintf("  Control-Before: %.4f (SD: %.4f)\n", cell_means[1,1], cell_sds[1,1]))
cat(sprintf("  Control-After: %.4f (SD: %.4f)\n", cell_means[1,2], cell_sds[1,2]))
cat(sprintf("  Treatment-Before: %.4f (SD: %.4f)\n", cell_means[2,1], cell_sds[2,1]))
cat(sprintf("  Treatment-After: %.4f (SD: %.4f)\n", cell_means[2,2], cell_sds[2,2]))

# Degrees of freedom breakdown
cat(sprintf("Degrees of freedom - Factor A: %d, Factor B: %d, Interaction: %d, Error: %d, Total: %d\n",
    factorA_df, factorB_df, interaction_df, residual_df, df_total))

cat("=== END R TWO-WAY ANOVA ===\n")